### DISCUSSION TITLE
Belief Propagation for Causal graphs

### DISCUSSION THREAD

**Discussion by h-sim-00** [pgmpy/pgmpy#1820] [Q&A]
Hi, 

I am working on causal discovery algorithms and have set up some code that results in an expression for an interventional query. For example, for the well-known front door criterion, my code outputs $$\sum_{Z}^{}P(Z|X=x)\sum_{x'}^{}P(y|z,x')P(x')$$ for the interventional query $$P(y|do(x))$$. Now, I want to use the belief propagation algorithm to obtain a numerical output for the above. I have been going through the documentation but cannot understand how to do that. I have my bayesian graph that I converted to the corresponding factor graph first and also calculated the CPDs required. 

Any help is appreciated!

**Comment by ankurankan** [comment]
@Coffee2000 Am I understanding correctly that you need the probability values for each term in $$\sum_{Z}^{}P(Z|X=x)\sum_{x'}^{}P(y|z,x')P(x')$$ ? Have you tried the `BeliefPropagation.query` method: https://pgmpy.org/exact_infer/bp.html#pgmpy.inference.ExactInference.BeliefPropagation.query . For example to compute the term $$P(Z | X=x)$$, you would need to do something like this:

```python
from pgmpy.inference import BeliefPropagation

model = some_bayesian_network
infer = BeliefPropagation(model)
infer.query(variables=['Z'], evidence={'X': x})
```

**Reply by h-sim-00** [reply]
Thank you for your response!

Yes, I tried this method exactly but I get different values when I compute the query by hand even for a simple dataset where all three (X, Y and Z)  are binary variables so I thought I am doing something wrong. 

How would I calculate the second term then? I assume there is a difference between evidence and conditions? 
Say for example for the second term I would need to calculate the following terms and sum all the values myself:
1. infer.query(variables=['Y'], evidence={'Z':0, 'X:0})
2. infer.query(variables=['Y'], evidence={'Z':1, 'X:0})
3. infer.query(variables=['Y'], evidence={'Z':0, 'X:1})
4. infer.query(variables=['Y'], evidence={'Z':1, 'X:1})
5. infer.query(variables=['X'])
Am I correct?

**Reply by ankurankan** [reply]
Assuming a model on 3 binary variables: X, Y, and Z. The state names would be 0 and 1 for each.
```python
from pgmpy.models import BayesianNetwork
from pgmpy.inference import VariableElimination

model = BayesianNetwork.get_random(
                n_nodes=3, 
                edge_prob=0.8,
                node_names=['X', 'Y', 'Z'], 
                n_states=2)
infer = VariableElimination(model)
```
The simplest way would be to write iterations exactly as the equation is. Some like the following (assuming that both the fixed X in $$P(Y|X=x)$$ and Y in $$P(y|z, x')$$ states are 1):
```python
In [43]: sum_outer = 0
    ...: for z in model.states['Z']:
    ...:     p_z = infer.query(['Z'], evidence={'X': 0}).get_value(Z=z)
    ...:     sum_inner = 0
    ...:     for x in model.states['X']:
    ...:         p_y_zx = infer.query(['Y'], evidence={'Z': z, 'X': x}).get_value(Y=0)
    ...:         p_x = infer.query(['X']).get_value(X=x)
    ...:         sum_inner += p_y_zx * p_x
    ...:     sum_outer += p_z * sum_inner

In [50]: sum_outer
Out[50]: np.float64(0.4442019376456231)
```
This would be the easiest way to write it. There should be ways to optimize it by running fewer queries by using factor operations such as multiplying whole distributions and marginalizing variables when doing summation.

**Reply by h-sim-00** [reply]
Yes, this method is the easiest but becomes intractable as the number of states and variables grows...that's the whole point of using Belief Propagation in the first place right? Can you please recommend a way to use the BP algorithm to compute the query or is it not possible to do so with the existing infrastructure? Apologies for asking again but I am a bit stumped on how to put it all together - a) I have a DAG  b) I can convert it to its corresponding factor graph and have calculated the required Conditional Probability Tables c) I also have the final interventional expression for my causal query d) Need the most optimal way to numerically compute said query (which is how I discovered the belief propagation method) 
I understand that the BP method is used to calculate marginals in an efficient way - so for my case calculating P(X), P(Y) and P(Z) but am wondering if this method can also calculate terms like $$\sum_{x'}^{}P(y|z,x')P(x')$$

Thank you!

**Reply by ankurankan** [reply]
Both the `BeliefPropagation.query` and `VariableElimination.query` methods are designed to compute distributions of the form $$P(X | \mathbf{Z}=\mathbf{z})$$. So, to use either of them for computing any term, you would have to express the term in that form. The example that I wrote was a naive way to do that, there should be other better ways to express that but would still need to reformulate the term.

If your only concern is intractability, I think even with the naive approach VariableElimination should be able to handle decently sized networks. The implementation of Variable Elimination is much more optimized compared to the Belief Propagation implementation.
